fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadReportingServiceImpl.java

```diff
@@ -151,21 +166,45 @@ public GenericResultsetData retrieveGenericResultset(final String name, final St
             final boolean isSelfServiceUserReport) {
 
         final long startTime = System.currentTimeMillis();
-        if (log.isDebugEnabled()) {
-            log.debug("STARTING REPORT: {}   Type: {}", LogParameterEscapeUtil.escapeLogParameter(name),
-                    LogParameterEscapeUtil.escapeLogParameter(type));
+        LOG.info("STARTING REPORT: {}   Type: {}", name, type);
+        StringBuilder sqlStringBuilder = new StringBuilder(200);
+        sqlStringBuilder.append(getSQLtoRun(name, type, queryParams, isSelfServiceUserReport));
+        final GenericResultsetData result;
+        boolean isPaginationAllowed = Boolean.parseBoolean(queryParams.get(ReportingConstants.IS_PAGINATION_ALLOWED));
+        Page<GenericResultsetData> reportData = this.paginationHelper.fetchPage(this.jdbcTemplate, sqlStringBuilder.toString(), null,
+                new ReportMapper(sqlStringBuilder));
+        if (isPaginationAllowed) {
+            sqlStringBuilder = retrieveGenericResultsetWithPagination(sqlStringBuilder, queryParams);
         }
+        result = this.genericDataService.fillGenericResultSet(sqlStringBuilder.toString());
+        int pageSize = this.configurationDomainService.reportsPaginationNumberOfItemsPerPage();
 
-        final String sql = getSQLtoRun(name, type, queryParams, isSelfServiceUserReport);
+        final long elapsed = System.currentTimeMillis() - startTime;
+        LOG.info("FINISHING Report/Request Name: {} - {}     Elapsed Time: {}", name, type, elapsed);
+        return GenericResultsetData.setTotalItemsAndRecordsPerPage(result, reportData.getTotalFilteredRecords(), pageSize);
+    }
 
-        final GenericResultsetData result = this.genericDataService.fillGenericResultSet(sql);
+    public StringBuilder retrieveGenericResultsetWithPagination(final StringBuilder sqlStringBuilder,
+            final Map<String, String> queryParams) {
+        retrieveGenericResultsetWithPaginationDataValidator(queryParams);
+        int pageSize = this.configurationDomainService.reportsPaginationNumberOfItemsPerPage();
+        int pageNo = Integer.parseInt(queryParams.get(ReportingConstants.OFFSET));
+
+        pageNo = pageNo * pageSize;
+        sqlStringBuilder.append(" order by ").append(queryParams.get(PAGINATION_ORDER_BY));
+        sqlStringBuilder.append(" ");
```

Comment:
Very very bad pattern. You're making this method to have side-effects while also returning the object. Decide on one and do it accordingly. Preferably return the Stirng.